### DISCUSSION TITLE
Secure external API endpoint

### DISCUSSION THREAD

**Discussion by nelsonsilva94** [pocketbase/pocketbase#1136] [Q&A] [answered]
First of all, big thanks to all the people involved in pocketbase!

I want to know if is possible and if yes, what is the best way to secure an external API endpoint.

Let's say I want to have some groups/roles, to which users will belong.

I then have a rust rocket backend and some endpoints that I want to secure and let only users from a certain group access them.

E.g. I have a user that belongs to a "supervisor" group, but not to an "analyst" group and in my rocket backend, I have an endpoint ".../xwy", which I only want users that belong to the "analyst" group to be able to access it and block it for users that don't belong to group "analyst".

Is there a chance to do that with pocketbase?

Sorry if it is a stupid question!

Thanks in advance

**Comment by ganigeorgiev** [comment, marked as answer]
This is difficult to generalize because it depends on how your app and data is structured and how you are planning to interact with PocketBase (only server-side, client-side or mixed).

PocketBase is stateless, aka. there are no sessions. A single app user is considered authenticated for a request if they send a valid `Authorization: TOKEN` header.

The auth token could be generated by sending `POST /api/collections/users/auth-with-password` request.

To verify in an external application that a token is valid, you can send a refresh API call with it (`POST /api/collections/users/auth-refresh`). If it fail, then the token is expired/invalid. If it succeed, then an updated token and user data will be returned.

More info on the available web APIs you could find in the "API Preview" from the Admin UI for each of your collections and/or check the general "Web APIs" docs - https://pocketbase.io/docs/api-records/.

**Reply by nelsonsilva94** [reply]
Thanks for the reply and clarifications!

I want to interact with PocketBase server-side only.

For now, I already have an authenticated user and a token, which I use to make API calls, I then created a custom field called "role" in the user collection (Is this the right approach?) and I want to know if I can use that user role information and in order to perform an API request to rocket.

**Reply by ganigeorgiev** [reply]
> Is this the right approach?
> I want to know if I can use that user role information and in order to perform an API request to rocket.

Yes, the custom user fields will be returned alongside the user data. So if you already have a token, you can send a `POST /api/collections/users/auth-refresh` and then based on the response you can determine whether to allow or reject access to your rocket api (the "role" custom field will be part of the user data).

**Reply by nelsonsilva94** [reply]
Thanks once again,

So that is all I need :)
